tests/test_knowledge_panels.py

```diff
@@ -296,3 +296,45 @@ def test_last_edits_kp_with_all_three_values(monkeypatch):
             ],
         }
     }
+
+
+def test_wikidata_kp(monkeypatch):
+    expected_url = "https://world.openfoodfacts.org/api/v2/taxonomy"
+    expected_kwargs = {
+        "params": {
+            "tagtype": "categories",
+            "fields": "wikidata",
+            "tags": "fr:fitou",
+        }
+    }
+    json_content = {"fr:fitou": {"parents": [], "wikidata": {"en": "Q470974"}}}
+    monkeypatch.setattr(
+        requests,
+        "get",
+        mock_get_factory(
+            expected_url,
+            expected_kwargs,
+            json_content,
+        ),
+    )
+    result = app.main.wikidata_kp(facet="category", value="fr:fitou")
```

Comment:
It would be better if we also monkeypatch the wikidata client…
This is not very difficult there again, you should be able to do it, in fact you just have to monkeypatch the "get" method of "wikidata.client.Client"
You monkey patch would return the current output of this function (you may put it in a separate json file if it's big)

Reply:
@alexgarel I'm confuse little bit this method `wikidata.client.Client` doesn't return any json it return a link like this `('https://www.wikidata.org/')`

Reply:
huum  (not sure to understand)

> `wikidata.client.Client` doesn't return any json 

Maybe when you call it without any argument, but your code will call it with a code, isn't it ?

Reply:
The lib handle all that search things for me , it return specific data according to entity.property

Reply:
Ok, so we are targeting those lines:

```python
    client = wikidata.client.Client()
    entity = client.get(entity_id)
```

The first line returns you an instance of the Client class, as expected:

```python
>>> client = wikidata.client.Client()
>>> type(client)
<class 'wikidata.client.Client'>
>>> client
wikidata.client.Client('https://www.wikidata.org/')
```

We don't need to change this, what we want to change is what we get on when we call client.get:

```
>>> client.get("Q47472153")
<wikidata.entity.Entity Q47472153>
```

When you use monkeypatch you can replace a function at module level, but you can also replace a method of a specific class.
So here the idea is to replace the "get" method (which we call in our code) with your fake implementation of the get.

Original get return an Entity instance which seems to have [a somewhat complicated implementation](https://github.com/dahlia/wikidata/blob/main/wikidata/entity.py#L138), so we can't create one in a simple mode in our code.
Instead you may simply return a namedtuple with the attributes your code use, and sample values in them.

Reply:
>  So here the idea is to replace the "get" method (which we call in our code) with your fake implementation of the get.

I forgot to point to an example. [In the documentation](https://docs.pytest.org/en/6.2.x/monkeypatch.html#simple-example-monkeypatching-functions) the first example monkey patch the Path class by changing its "home" method.
You want to do the same: monkey patch the wikidata.client.Client class by changing its "get" method.

Reply:
You misunderstood me @sumit-158 we don't need the test_external_lib.py test (I think you were just using it to experiment).

We need to monkeypatch the Client.get in this test. 
This is why you will need the FakeEntity class.

That said putting the mocking machinary in a separate file (wikidata_utils.py) is a good idea :-)

Reply:
Try again, and if it's really too hard, I will give an help !